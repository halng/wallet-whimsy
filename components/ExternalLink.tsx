/*
* Copyright 2025 Hal Ng
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import { Link } from 'expo-router';
import { openBrowserAsync } from 'expo-web-browser';
import { type ComponentProps } from 'react';
import { Platform } from 'react-native';

type Props = Omit<ComponentProps<typeof Link>, 'href'> & { href: string };

export function ExternalLink({ href, ...rest }: Props) {
  return (
    <Link
      target="_blank"
      {...rest}
      href={href}
      onPress={async (event) => {
        if (Platform.OS !== 'web') {
          // Prevent the default behavior of linking to the default browser on native.
          event.preventDefault();
          // Open the link in an in-app browser.
          await openBrowserAsync(href);
        }
      }}
    />
  );
}
