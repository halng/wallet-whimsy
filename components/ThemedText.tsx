/*
* Copyright 2025 Hal Ng
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import { Text, type TextProps, StyleSheet } from 'react-native';

import { useThemeColor } from '@/hooks/useThemeColor';

export type ThemedTextProps = TextProps & {
  lightColor?: string;
  darkColor?: string;
  type?: 'default' | 'title' | 'defaultSemiBold' | 'subtitle' | 'link';
};

export function ThemedText({
  style,
  lightColor,
  darkColor,
  type = 'default',
  ...rest
}: ThemedTextProps) {
  const color = useThemeColor({ light: lightColor, dark: darkColor }, 'text');

  return (
    <Text
      style={[
        { color },
        type === 'default' ? styles.default : undefined,
        type === 'title' ? styles.title : undefined,
        type === 'defaultSemiBold' ? styles.defaultSemiBold : undefined,
        type === 'subtitle' ? styles.subtitle : undefined,
        type === 'link' ? styles.link : undefined,
        style,
      ]}
      {...rest}
    />
  );
}

const styles = StyleSheet.create({
  default: {
    fontSize: 16,
    lineHeight: 24,
  },
  defaultSemiBold: {
    fontSize: 16,
    lineHeight: 24,
    fontWeight: '600',
  },
  title: {
    fontSize: 32,
    fontWeight: 'bold',
    lineHeight: 32,
  },
  subtitle: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  link: {
    lineHeight: 30,
    fontSize: 16,
    color: '#0a7ea4',
  },
});
