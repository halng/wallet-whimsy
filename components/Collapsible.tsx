/*
 * Copyright 2025 Hal Ng
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { PropsWithChildren, useState } from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';

import { ThemedText } from '@/components/ThemedText';
import { ThemedView } from '@/components/ThemedView';
import { IconSymbol } from '@/components/ui/IconSymbol';
import { Colors } from '@/constants/Colors';
import { useColorScheme } from '@/hooks/useColorScheme';

export function Collapsible({ children, title }: PropsWithChildren & { title: string }) {
  const [isOpen, setIsOpen] = useState(false);
  const theme = useColorScheme() ?? 'light';

  return (
    <ThemedView>
      <TouchableOpacity
        style={styles.heading}
        onPress={() => setIsOpen(value => !value)}
        activeOpacity={0.8}
      >
        <IconSymbol
          name="chevron.right"
          size={18}
          weight="medium"
          color={theme === 'light' ? Colors.light.icon : Colors.dark.icon}
          style={{ transform: [{ rotate: isOpen ? '90deg' : '0deg' }] }}
        />

        <ThemedText type="defaultSemiBold">{title}</ThemedText>
      </TouchableOpacity>
      {isOpen && <ThemedView style={styles.content}>{children}</ThemedView>}
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  heading: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 6,
  },
  content: {
    marginTop: 6,
    marginLeft: 24,
  },
});
