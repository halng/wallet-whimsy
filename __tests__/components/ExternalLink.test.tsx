/*
 * Copyright 2025 Hal Ng
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { render, screen, fireEvent } from '@testing-library/react-native';
import { Platform } from 'react-native';
import { openBrowserAsync } from 'expo-web-browser';

import { ExternalLink } from '@/components/ExternalLink';

// Mock the dependencies
jest.mock('expo-web-browser', () => ({
  openBrowserAsync: jest.fn(),
}));
describe('ExternalLink Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    Platform.OS = 'web';
  });

  // Mock event object
  const mockEvent = {
    preventDefault: jest.fn(),
  };

  it('renders correctly with required props', () => {
    render(
      <ExternalLink href="https://example.com" testID="external-link">
        Test Link
      </ExternalLink>,
    );

    const link = screen.getByTestId('external-link');
    expect(link).toBeTruthy();
  });

  it('renders with correct text content', () => {
    render(
      <ExternalLink href="https://example.com" testID="external-link">
        Test Link
      </ExternalLink>,
    );

    const link = screen.getByText('Test Link');
    expect(link).toBeTruthy();
  });

  it('does not call openBrowserAsync on web platform', async () => {
    Platform.OS = 'web';
    render(
      <ExternalLink href="https://example.com" testID="external-link">
        Test Link
      </ExternalLink>,
    );

    const link = screen.getByTestId('external-link');
    fireEvent.press(link, mockEvent);

    expect(openBrowserAsync).not.toHaveBeenCalled();
    expect(mockEvent.preventDefault).not.toHaveBeenCalled();
  });

  it('calls openBrowserAsync on iOS platform', async () => {
    Platform.OS = 'ios';
    render(
      <ExternalLink href="https://example.com" testID="external-link">
        Test Link
      </ExternalLink>,
    );

    const link = screen.getByTestId('external-link');
    fireEvent.press(link, mockEvent);

    expect(mockEvent.preventDefault).toHaveBeenCalled();
    expect(openBrowserAsync).toHaveBeenCalledWith('https://example.com');
  });

  it('calls openBrowserAsync on Android platform', async () => {
    Platform.OS = 'android';
    render(
      <ExternalLink href="https://example.com" testID="external-link">
        Test Link
      </ExternalLink>,
    );

    const link = screen.getByTestId('external-link');
    fireEvent.press(link, mockEvent);

    expect(mockEvent.preventDefault).toHaveBeenCalled();
    expect(openBrowserAsync).toHaveBeenCalledWith('https://example.com');
  });

  it('passes through additional props', () => {
    render(
      <ExternalLink
        href="https://example.com"
        testID="external-link"
        accessibilityLabel="Accessible link"
      >
        Test Link
      </ExternalLink>,
    );

    const link = screen.getByTestId('external-link');
    expect(link).toBeTruthy();
  });
});
