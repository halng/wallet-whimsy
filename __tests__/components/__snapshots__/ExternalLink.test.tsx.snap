// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - ExternalLink Snapshot test 1`] = `
<Text
  href="/"
  onPress={[Function]}
  role="link"
/>
`;
